e2e/tests/upgrade-test.rs

```diff
@@ -1,36 +1,438 @@
+    assert!(from_timescaledb_version <= to_timescaledb_version);
+
+    // install the promscale extension at the "from_version"
+    install_promscale_ext(&mut from_client, &from_version);
+
+    // load test data if configured to
+    if with_data {
+        load_data(&from_container);
+    }
+
+    // checkpoint
+    from_client
+        .execute("checkpoint", &[])
+        .expect("failed to checkpoint");
+
+    // shut it down
+    from_client
+        .close()
+        .expect("failed to close database connection");
+    from_container.stop();
+
+    /**********************************************************************************************/
+    // UPGRADE TO
+    info!("UPGRADE TO");
+    let to_blueprint = PostgresContainerBlueprint::new()
+        .with_image_uri(to_image_uri)
+        .with_volume(script_dir, "/scripts")
+        .with_volume(data_dir, volume_map_path)
+        .with_env_var("PGDATA", "/var/lib/postgresql/data")
+        .with_db("db")
+        .with_user("postgres");
+    let to_container = to_blueprint.run();
+    info!("to_container id: {}", to_container.id());
+
+    // upgrade the timescaledb extension if we need to
+    if from_timescaledb_version != to_timescaledb_version {
+        info!(
+            "upgrading from timescaledb {} to {}",
+            from_timescaledb_version, to_timescaledb_version
+        );
+        update_extension(
+            to_container.id(),
+            "postgres",
+            "db",
+            "timescaledb",
+            &to_timescaledb_version,
+        );
+    }
+
+    // update the promscale extension
+    update_extension(
+        to_container.id(),
+        "postgres",
+        "db",
+        "promscale",
+        &to_version,
+    );
+
+    // snapshot the database
+    let upgraded_snapshot = snapshot_db(
+        to_container.id(),
+        "db",
+        "postgres",
+        working_dir.join("upgraded-snapshot.txt").as_path(),
+    );
+
+    // shut it down
+    to_container.stop();
+
+    // compare the snapshots
+    let are_equal = are_snapshots_equal(baseline_snapshot, upgraded_snapshot);
+    assert!(are_equal);
+}
+
+fn drop_toolkit_extension(client: &mut Client) {
+    // Some docker images may also have timescaledb_toolkit installed in a
+    // template, which messes with these tests because toolkit was not always
+    // present. We can just drop it, it's not needed for the upgrade tests.
+    client
+        .execute("DROP EXTENSION IF EXISTS timescaledb_toolkit;", &[])
+        .expect("failed to drop the toolkit extension");
+}
+
+fn from_image(flavor: &Flavor, pg_version: &PgVersion, from_version: &FromVersion) -> String {
+    match flavor {
+        Flavor::Alpine => match from_version {
+            // Our timescaledev images are deprecated, but they are the best option for an alpine
+            // image at the moment. They also are nicely tagged with the exact extension version.
```

Comment:
Oh gosh. This sounds like a pain. Perhaps we can load this map from a file. I would lump alpine versions (from the previous comment) into the same file then.

Reply:
Doesn't have to be a text file: it could be a Rust module, but in a separate file with nothing else but image name picking logic.

Reply:
I've done an initial pass at this. Still needs some documentation.

Reply:
Much appreciated 👍🏼